### PR TITLE
fix flaky testPublish_Object

### PR DESCRIPTION
**Pull request by kateryna216**
### What is the purpose of this PR
This PR fixes the flaky test com.google.cloud.spring.pubsub.core.PubSubTemplateTests.testPublish_Object.

### Why the test fails
It fails because protobuf does not guarantee serialization/deserialization order of fields.

### Reproduce the test failure
Run the test with 'NonDex' maven plugin and the command
`mvn -pl spring-cloud-gcp-pubsub edu.illinois:nondex-maven-plugin:1.1.2:nondex -Dtest=com.google.cloud.spring.pubsub.core.PubSubTemplateTests#testPublish_Object`

### Error
```
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.google.cloud.spring.pubsub.core.PubSubTemplateTests
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.997 s <<< FAILURE! - in com.google.cloud.spring.pubsub.core.PubSubTemplateTests
[ERROR] testPublish_Object  Time elapsed: 1.964 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: 

expected: "{"@class":"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload","name":"allowed","value":12345}"
 but was: "{"@class":"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload","value":12345,"name":"allowed"}"
	at com.google.cloud.spring.pubsub.core.PubSubTemplateTests.lambda$testPublish_Object$0(PubSubTemplateTests.java:136)
	at com.google.cloud.spring.pubsub.core.PubSubTemplateTests.testPublish_Object(PubSubTemplateTests.java:145)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PubSubTemplateTests.testPublish_Object:145->lambda$testPublish_Object$0:136 
expected: "{"@class":"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload","name":"allowed","value":12345}"
 but was: "{"@class":"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload","value":12345,"name":"allowed"}"
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```
### Fix
Handle different orders when comparing expected with actual.

### CODE REVIEW

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/PubSubTemplateTests.java [hunk 1]
```diff
@@ -132,15 +133,18 @@ void testPublish_Object() throws IOException {
...
+              Arrays.sort(expected);
+              Arrays.sort(actual);
+
+              assertThat(actual).isEqualTo(expected);
               return null;
             })
-        .when(pubSubPublisherTemplate)
-        .publish(eq("test"), any());
+            .when(pubSubPublisherTemplate)
+            .publish(eq("test"), any());
```
Comment (winglam) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/PubSubTemplateTests.java]: Can you undo these unnecessary changes?

File: spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/PubSubTemplateTests.java [hunk 2]
```diff
@@ -132,15 +133,18 @@ void testPublish_Object() throws IOException {
...
-                  .isEqualTo(
-                      "{\"@class\":"
-                          + "\"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload\""
-                          + ",\"name\":\"allowed\",\"value\":12345}");
+              char[] expected = ("{\"@class\":"
+                      + "\"com.google.cloud.spring.pubsub.core.test.allowed.AllowedPayload\""
+                      + ",\"name\":\"allowed\",\"value\":12345}").toCharArray();
+              char[] actual = message.getData().toStringUtf8().toCharArray();
+              Arrays.sort(expected);
+              Arrays.sort(actual);
```
Comment (winglam) [inline spring-cloud-gcp-pubsub/src/test/java/com/google/cloud/spring/pubsub/core/PubSubTemplateTests.java]: Another way to have fixed this test could have been to check that the size of the two arrays are the same and then just check whether `actual` contains all of the elements that are in `expected`. Such a fix would add more lines of code, but would also be more efficient than your fix. 

In any case, the developers may not care about this difference though, so you can keep your current fix as is.

**Comment by winglam** [comment]
Can you please prepare the description of the pull request to be similar to the following?
https://github.com/apache/commons-collections/pull/336

**Comment by winglam** [comment]
Your changes to improve the code looks good.

As I mentioned before, can you please improve the description of the pull request to be similar to the following?
https://github.com/apache/commons-collections/pull/336

**Comment by kateryna216** [comment]
this PR got closed because I had to change the email in old commit to verify my Google email to contribute to Google project but it all got screwed

**Comment by winglam** [comment]
To increase the chances that your PR will be accepted, a better explanation for why the test fails is the following.
Line 145 of PubSubTemplateTests.java starts a call that eventually serializes some data into JSON format. This data is then deserialized in the callback (Line 132). The order in which JSON data is deserialized is not guaranteed, yet the original assertion of the test assumes a deterministic order (Line 135). This pull request aims to not check that the string is an exact match, but instead that the proper entries are in the string.

Please also add the following to "Reproduce the test failure".
Fixing the flaky test now may prevent flaky test failures in future versions.

Please feel free to directly open your PR to the actual repository after you make the above changes. Be sure to keep an eye on communication from the developers for your PR and keep me updated on any changes to your PR. Best of luck to you for getting this PR accepted!

**Comment by winglam** [comment]
Actually one other small comment before you make the PR, please look over the following StackOverflow post and squash all of your commits in this PR into one commit before you open your PR to the developer's repo.
https://stackoverflow.com/questions/5189560/how-do-i-squash-my-last-n-commits-together

One commit in your PR is much cleaner than the many commits currently in this PR.
